Implement vector finite elements

lindsayad opened the pull request:
Important changes:
  - Introduced a finite element variable base class called MooseVariableFE
    - Contains methods and data that don't deal with the interpolated solution (e.g.
      _u, _grad_u, _curl_u, etc.)
  - MooseVariableField (this name could be changed) is a class template that
    inherits from MooseVariableFE and implements the methods and data that deal
    with the interpolated solution data members.
    - I've typedef'd `MooseVariableField<Real>` to be the well-known MooseVariable
    - `MooseVariableField<VectorValue<Real>>` is typedef'd as MooseVariableVector
  - Some methods in Assembly have been templated so that they return the proper shape
    functions corresponding to scalar shape functions for MooseVariable and vector
    shape functions for MooseVariableVector
  - Refactored EigenKernel to inherit directly from Kernel instead of KernelBase
  - VectorKernel inherits from KernelBase
  - Created a new BoundaryCondition class called IntegratedBCBase
    - Traditional IntegratedBC is the scalar version that inherits from IntegratedBCBase
    - VectorIntegratedBC is the vector version that inherits from IntegratedBCBase
  - Added a test that mimics libmesh/examples/vector_fe/ex3
  - Implements `NEDELEC_ONE` but can very easily add `LAGRANGE_VEC` which at least @dschwen is interested in

There are undoubtedly some rough edges in here, and I should add a test for coupling a vector variable with a regular variable for example, but I figured I'd get this up here in case of discussion.

Closes (#10049)

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/132998/) on 29513b9 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 08bb17a1b082a440272069e3332241e8c225f840`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/10238/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/10238/style.patch | git apply -v`

lindsayad commented:
The test mimics example 3 in libmesh's `examples/vector_fe` directory. Results:

### ux
<img width="794" alt="ux" src="https://user-images.githubusercontent.com/10248304/32665567-f870fae6-c5f1-11e7-9db9-da6372bcb443.png">

### uy
<img width="719" alt="uy" src="https://user-images.githubusercontent.com/10248304/32665573-ff4121ca-c5f1-11e7-9f13-7f3a4544142e.png">

tagging @cticenhour

dschwen commented:
```
MooseVariable.h: No such file or directory
```

Breaks a ton of builds (Marmot for example)
Can you maybe add that file back in (even if it now just contains a single(?) include directive to include the new file)?

lindsayad commented:
I like that fix. Will do

> On Nov 10, 2017, at 11:40 AM, Daniel Schwen <notifications@github.com> wrote:
> 
> MooseVariable.h: No such file or directory
> Breaks a ton of builds (Marmot for example)
> Can you maybe add that file back in (even if it now just contains a single(?) include directive to include the new file)?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/133046/) on e5694d7 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 7276caca9ea6dd7a92a5e5293c8f784953f3a649`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/10238/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/10238/style.patch | git apply -v`

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 7f3ced2710c67ef3f7d7ba512090259b09997c74
+Subproject commit 4d08770f7f11bb9bfb88cb631fdd17165207dad5
```
**Caution!** This contains a submodule update

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/133121/) on 6bcbf71 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 9d00e49fd410c275fe41c47e387c89d5a16a34d3`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/10238/style.patch).
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10238/style.patch | git apply -v`

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/133137/) on 5108786 wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/10238/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Modules debug](https://www.moosebuild.org/job/134294/) on 31d3337 : invalidated by @lindsayad

At least one of these time-out jobs runs at the same speed for me on devel and the feature branch...

gridley commented:
How far do you think this is from completion @lindsayad? I may go ahead and pull your branch to start P1 in Moltres. Also, is there a _div_u available? Just about to look at the code, but so many files appear to have been changed; it's hard to pinpoint where to look.

andrsd commented:
> Also, is there a _div_u available?

I searched through the patch and found only `_curl_u`.

andrsd commented:
I plan on looking at this PR after the holidays.

lindsayad commented:
David's right; there's no div_u in this patch, but it would not be
difficult to add if it's desired.

On Tue, Dec 19, 2017 at 8:06 AM, David Andrs <notifications@github.com>
wrote:

> I plan on looking at this PR after the holidays.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/10238#issuecomment-352785299>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AJxgcGiGcxhGUB3nEVhdnZUxhutNHCpVks5tB9DjgaJpZM4QZtul>
> .
>

lindsayad commented:
![inherit-graph](https://user-images.githubusercontent.com/10248304/34731026-472dad04-f51e-11e7-8ee3-0fcba880c052.png)

@friedmud Here's the inheritance graph for moose variables for this PR. I know the names aren't great, and I am open to suggestions for better. One of the reasons for having the `MooseVariableFE` class is so that I can have containers, e.g. vectors, that contain both scalar and vector finite element variables. The inheritance is a bit similar to that of the [FE libmesh classes](https://libmesh.github.io/doxygen/classlibMesh_1_1FEGenericBase.html) where `FEAbstract` is akin to `MooseVariableFE`, and `FEGenericBase<OutputType>` is akin to `MooseVariableField<OutputType>`

andrsd commented on framework/src/outputs/Exodus.C:
```diff
@@ -195,6 +195,21 @@ Exodus::outputSetup()
+Exodus::outputVectorVariables()
+{
+  // std::vector<std::string> vector(getVectorVariableOutput().begin(),
+  // getVectorVariableOutput().end());
```
we do not need this...

andrsd commented on framework/src/postprocessors/ElementIntegralVariablePostprocessor.C:
```diff
@@ -26,12 +26,18 @@ validParams<ElementIntegralVariablePostprocessor>()
+    // interface level...
+    variable_name = parameters.getVecMooseType("variable")[0];
+
+  addMooseVariableDependency(&_subproblem.getVariable(_tid, variable_name));
```
These bits that repeat feels like they should have stayed in `MooseVariableInterface` you removed.  Maybe we need `MooseVectorVariableInterface` for the vector-valued systems?

andrsd commented on modules/navier_stokes/src/userobjects/CNSFVGreenGaussSlopeReconstruction.C:
```diff
@@ -35,11 +35,11 @@ validParams<CNSFVGreenGaussSlopeReconstruction>()
+    _rhou(dynamic_cast<MooseVariable *>(getVar("rhou", 0))),
+    _rhov(isCoupled("rhov") ? dynamic_cast<MooseVariable *>(getVar("rhov", 0)) : nullptr),
+    _rhow(isCoupled("rhow") ? dynamic_cast<MooseVariable *>(getVar("rhow", 0)) : nullptr),
+    _rhoe(dynamic_cast<MooseVariable *>(getVar("rhoe", 0))),
```
Yuck!

andrsd commented on modules/phase_field/include/kernels/CHInterfaceBase.h:
```diff
@@ -67,9 +67,9 @@ CHInterfaceBase<T>::CHInterfaceBase(const InputParameters & parameters)
-    _second_phi(secondPhi()),
+    _second_u(_var.secondSln()),
+    _second_test(_var.secondPhi()),
+    _second_phi(_var.secondPhi()),
```
And there is another reason why `MooseVariableInterface` should have stayed...

lindsayad replied:
But it's only 5 more chracters vs. a whole intermediate class? Now the repeated call of `_sys.getVariable(_tid, parameters.get<NonlinearVariableName>("variable"))` (or `_subproblem...`) in all the computing object classes does seem tedious I'll agree

andrsd replied:
And let's not forget we might hide some smart things in that `MooseVariableInterface` c-tor. Those would get propagated into all interfaces that are using the class...

andrsd commented on test/tests/ics/dependency/monomial.i:
```diff
@@ -86,32 +86,36 @@
     variable = v
     boundary = right
     value = 1
+    penalty = 1000
```
These changes look unrelated to your PR. Why was this changed?

lindsayad replied:
With my changes this test errors out. I suppose there are a couple of ways to think about DirichletBC. If we think about it as assigning values for DOFs then I suppose my changes should be compatible with this test as it stands in HEAD. However, if we think about them as assigning values on domain boundaries then I don't think using a DirichletBC makes sense in the context of discontinuous monomials.

andrsd commented on test/tests/kernels/vector_fe/tests:
```diff
@@ -0,0 +1,8 @@
+    input = 'vector_kernel.i'
+    exodiff = 'vector_kernel_out.e'
+    abs_zero = 1e-9
+  [../]
```
One test rules them all? :-)

lindsayad replied:
I've added a couple more tests. One demonstrates coupling a vector variable into a standard kernel. The other tests the Jacobian for the former case since we introduce a `_vector_phi` member on top of `_phi`.

I would like to add a LAGRANGE_VEC test, but without too much more work it will have to use integrated boundary conditions. It's currently possible in libmesh to specify dirichlet boundary conditions for LAGRANGE_VEC, but we'll have to add some additional machinery to MOOSE to have it work for us.

lindsayad commented:
Thanks a lot for the review @andrsd. I hope to get to it soon.

lindsayad commented on framework/src/bcs/IntegratedBC.C:
```diff
@@ -27,65 +27,28 @@ template <>
-    _current_side_volume(_assembly.sideElemVolume()),
-
+  : IntegratedBCBase(parameters), // False is because this is NOT nodal
+    _var(dynamic_cast<MooseVariable &>(_base_var)),
```
@andrsd Would templating the getVariable methods make sense?

andrsd replied:
Yes, it would.

andrsd replied:
Although you might have an issue in `AuxiliarySystem`, because your PR does not allow vector-valued aux-variables. But, you should be able to use explicit instantiation to prevent people to add/get them.

lindsayad replied:
What design do you like better: 
1) A single container of MooseVariableBase pointers in the VariableWarehouse with `dynamic_cast` checking in the templated `SystemBase` getter
2) Separate (unordered)maps of MooseVariables and MooseVariableVectors and a templated getter method in the VariableWarehouse that removes the need for dynamic casting in the `SystemBase` getter

andrsd replied:
I'd go with 2. We already have normal + scalar variables there, so it is just logical we add vector valued variables there.

lindsayad committed:
Make different methods in coupleable for getting standard vs vector variables.

lindsayad committed:
Change VariableWarehouse variables() method to fieldVariables()

lindsayad committed:
Restore ability to do LAGRANGE_VEC for any dimension.

lindsayad committed:
Rename MooseVariableVector to VectorMooseVariable

lindsayad committed:
Create templated fieldVariable warehouse getter and create separate maps for scalar and vector field variables.

lindsayad committed:
Implement vector finite elements

Important changes:
  - Introduced a finite element variable base class called MooseVariableFE
    - Contains methods and data that don't deal with the interpolated solution (e.g.
      _u, _grad_u, _curl_u, etc.)
  - MooseVariableField (this name could be changed) is a class template that
    inherits from MooseVariableFE and implements the methods and data that deal
    with the interpolated solution data members.
    - I've typedef'd MooseVariableField<Real> to be the well-known MooseVariable
    - MooseVariableField<VectorValue<Real>> is typedef'd as MooseVariableVector
  - Some methods in Assembly have been templated so that they return the proper shape
    functions corresponding to scalar shape functions for MooseVariable and vector
    shape functions for MooseVariableVector
  - Refactored EigenKernel to inherit directly from Kernel instead of KernelBase
  - VectorKernel inherits from KernelBase
  - Created a new BoundaryCondition class called IntegratedBCBase
    - Traditional IntegratedBC is the scalar version that inherits from IntegratedBCBase
    - VectorIntegratedBC is the vector version that inherits from IntegratedBCBase
  - Add methods in FEProblemBase for standard and vector variable getting
    - This is in addition to the getVariable method which returns a pointer to the abstract
      base class MooseVariableFE
  - Split phi() accessor methods in Assembly into writeable and const read-only
  - Overload phi() accessors for different variable types. Saves some lines of code
    - Maintain some zero-arg accessor methods for backwards compatibility with the ShapeUserObject
      that doesn't have a MooseVariable* object with which to call the correct 1-arg methods
  - Template getter method in SystemBase
  - Added a test that mimics libmesh/examples/vector_fe/ex3

Closes (#10049)

lindsayad committed:
Template the *VariableInterface*s.

lindsayad commented:
@andrsd Hopefully the commit history looks a bit better now. Thanks for pushing me to explore `fixup` and `autosquash`. Neat tools.

Next thing I'll work on is adding a test demonstrating use of `LAGRANGE_VEC`

lindsayad committed:
Add additional tests

Tests include:
  - Coupling example between a scalar charge affected by the vector field
  - Jacobian test which checks contribution to off-diagonal jacobian coming from
    the vector variable (have to create _vector_phi variable)
  - LAGRANGE_VEC Poisson example that emulates the libmesh vector_fe example 1

Additional change:
  - Modify PetscJacobianTester to allow comparison against Petsc constant state
    1.0 or constant state -1.0. This is necessary in our current case because
    we don't yet have projection methods for NEDELEC basis functions and hence
    cannot specify non-zero initial conditions

lindsayad commented:
I think this is ready for further review

andrsd commented on framework/include/utils/MooseTypes.h:
```diff
@@ -71,6 +72,21 @@
+namespace libMesh
+{
+template <typename>
+class VectorValue;
```
Are you not getting this via `#include "libmesh/vector_value.h"`?

andrsd commented on framework/include/utils/MooseTypes.h:
```diff
@@ -71,6 +72,21 @@
+template <typename>
+class TensorValue;
+template <unsigned int, typename>
+class TypeNTensor;
```
`#include "libmesh/tensor_value.h"` and `"type_n_tensor.h"` and then you do not need this.

andrsd commented on framework/src/bcs/IntegratedBCBase.C:
```diff
@@ -0,0 +1,66 @@
+
+#include "IntegratedBCBase.h"
+
+// MOOSE includes
```
The first one is not MOOSE include? ;-)
Let's not put cpt. Obvious in the code.

andrsd commented on framework/src/kernels/VectorKernel.C:
```diff
@@ -0,0 +1,132 @@
+
+#include "VectorKernel.h"
+
+// MOOSE includes
```
cpt. Obvious.

andrsd commented on modules/phase_field/include/kernels/CahnHilliardBase.h:
```diff
@@ -100,7 +100,7 @@ CahnHilliardBase<T>::CahnHilliardBase(const InputParameters & parameters)
     }
 
-    _grad_vars[i + 1] = &(_coupled_moose_vars[i]->gradSln());
+    _grad_vars[i + 1] = &this->_sys.template getFieldVariable<Real>(this->_tid, iname).gradSln();
```
I do not think we want this in user code. Can we provide better API?

andrsd commented on modules/phase_field/src/kernels/ACInterface.C:
```diff
@@ -41,7 +41,7 @@ ACInterface::ACInterface(const InputParameters & parameters)
   for (unsigned int i = 0; i < _nvar; ++i)
   {
-    MooseVariable * ivar = _coupled_moose_vars[i];
+    MooseVariable * ivar = dynamic_cast<MooseVariable *>(_coupled_moose_vars[i]);
```
in `ACSwitching` you used `MooseVariableFE` and did not have to use `dynamic_cast`. why could not you use it here?

lindsayad replied:
We call the `gradSln` method on `ivar` which isn't implemented for the abstract `MooseVariableFE`. We could do a dance where we get the `MooseVariable` using the `_subproblem` or `_sys` getters with `_coupled_moose_vars[i]->name()`. Or I could add another method(s) in `Coupleable` like `getCoupledStandardMooseVars` (and `getCoupledVectorMooseVars`) which would require adding a few more data members as well. I'm feeling lazy but will implement whatever design you think best

andrsd commented on test/include/kernels/EFieldAdvection.h:
```diff
@@ -0,0 +1,43 @@
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+#ifndef EFIELDADVECTION_H_
+#define EFIELDADVECTION_H_
```
no trailing `_`

andrsd commented on test/include/kernels/EFieldAdvection.h:
```diff
@@ -0,0 +1,43 @@
+  virtual Real computeQpJacobian();
+  virtual Real computeQpOffDiagJacobian(unsigned int jvar);
+
+private:
```
So whoever inherits from this won't be able to use those?

andrsd commented on framework/src/base/NeighborCoupleable.C:
```diff
@@ -32,9 +32,9 @@ NeighborCoupleable::~NeighborCoupleable() {}
 NeighborCoupleable::coupledNeighborValue(const std::string & var_name, unsigned int comp)
 {
-  MooseVariable * var = getVar(var_name, comp);
+  MooseVariable * var = dynamic_cast<MooseVariable *>(getVar(var_name, comp));
```
Are we sure this will be always not NULL?

andrsd replied:
Can we have an API that will give back `MooseVariable *` that will be checked for `!nullptr` and in case if `nullptr`, we get an error message. It would be a lot better than to have not-checked `dynamic_cast`s.

lindsayad replied:
`Coupleable::getVar` does this but these eluded my search and destroy of unnecessary casts

andrsd replied:
Right, I missed this was in `NeighborCoupleable`. Easy fix then. 👍

lindsayad commented on framework/include/kernels/KernelBase.h:
```diff
@@ -93,13 +96,26 @@ class KernelBase : public MooseObject,
   virtual bool isEigenKernel() const { return _eigen_kernel; }
 
+protected:
+  /// Compute this Kernel's contribution to the residual at the current quadrature point
```
I thought our style was to use triple slash for doxygen comments that fit on a single line and the regular block style for multi-line comments?

andrsd replied:
IIRC, `///` for variables and `/** */` for methods.

lindsayad replied:
Ah, got ya

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/153854/) on 8c91a0e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.org/docs/PRs/10238/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10238/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7355abbe7cd56969185b85ab49b2a01c9926b8ee`

lindsayad committed:
Add additional containers and accessor methods in Coupleable for standard and vector moose vars.

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/153956/) on e22f113 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.org/docs/PRs/10238/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10238/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7355abbe7cd56969185b85ab49b2a01c9926b8ee`

moosebuild commented:
Job [Test Intel](https://www.moosebuild.org/job/153982/) on 8807cac : invalidated by @lindsayad

This PR shouldn&#39;t have touched this failing test

moosebuild commented:
Job [Test Tiamat](https://www.moosebuild.org/job/154175/) on 8b004ad : invalidated by @lindsayad

Tiamat fix in

moosebuild commented:
Job [Test Tiamat](https://www.moosebuild.org/job/154175/) on 8b004ad : invalidated by @lindsayad

lindsayad commented:
If we're happy with what's going on here, I will start preparing patches for the Apps...

mangerij commented:
So when can we do light scattering? :)

On Mon, Feb 12, 2018 at 7:09 PM Alex Lindsay <notifications@github.com>
wrote:

> If we're happy with what's going on here, I will start preparing patches
> for the Apps...
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/10238#issuecomment-365010037>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHPulGCLdijG7kENwzKe5hxWDH6SmJ_Gks5tUH5igaJpZM4QZtul>
> .
>

lindsayad committed:
Add documentation

lindsayad commented:
So, I have patches ready for all the failing apps, built, tested, and submitted to each. So unless there's anything else we want changed, I'm considering this ready to go. @andrsd your thoughts?

andrsd commented:
> @andrsd your thoughts?

Let the world explode!

dschwen commented:
@lindsayad, is the divergence not just the trace of the VectorVariableGradient?

lindsayad commented:
In Cartesian coordinates this is true but not for general coordinate systems IIRC

> On Mar 15, 2018, at 8:24 PM, Daniel Schwen <notifications@github.com> wrote:
> 
> @lindsayad, is the divergence not just the trace of the VectorVariableGradient?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

dschwen commented:
So, has anyone already tried implementing Maxwell's equations then? 🏓 @tonkmr

mangerij commented:
I have a few projects in mind but at the moment it is very hard to find the
time.

Would be cool to see how the framework handles light scattering problems
(those typically calculated from FDTD methods). I imagine the materials
system could be useful in defining things like the conductivity and
dielectric tensors.



On Fri, Mar 16, 2018 at 4:56 AM, Daniel Schwen <notifications@github.com>
wrote:

> So, has anyone already tried implementing Maxwell's equations then? 🏓
> @tonkmr <https://github.com/tonkmr>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/10238#issuecomment-373596393>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHPulFeWAGsVXKO9Ijpnb7BIG1QiX38iks5tezgDgaJpZM4QZtul>
> .
>

tonkmr commented:
I am currently working on a proposal where, if it gets funded, we will start to solve Maxwell's equations

lindsayad commented:
You often make a sub-selection of maxwells equations you want to solve. The first vector fe test in the framework essentially solves a combination of amperes and faradays laws (combined into a single curl curl equation). Or you can use alternative gauge formulations which use potentials as opposed to the standard form of maxwells equations. There's a new test in #10933 that solves a Coulomb gauge equation, although that formulation really isn't used very often in practice. 

> On Mar 16, 2018, at 6:54 AM, Michael Tonks <notifications@github.com> wrote:
> 
> I am currently working on a proposal where, if it gets funded, we will start to solve Maxwell's equations
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

mangerij commented:
Solving Helmholtz equations with anisotropic inhomogeneous dielectric
fields would be ideal. We could even load them in from Ferret.

I don't know if it's even doable. I think its a nightmare of numerics to
say your wave can arbitrarily change speed AND polarization as it passes
from one grid point to the next.

On Fri, Mar 16, 2018 at 5:02 PM, Alex Lindsay <notifications@github.com>
wrote:

> You often make a sub-selection of maxwells equations you want to solve.
> The first vector fe test in the framework essentially solves a combination
> of amperes and faradays laws (combined into a single curl curl equation).
> Or you can use alternative gauge formulations which use potentials as
> opposed to the standard form of maxwells equations. There's a new test in
> #10933 that solves a Coulomb gauge equation, although that formulation
> really isn't used very often in practice.
>
> > On Mar 16, 2018, at 6:54 AM, Michael Tonks <notifications@github.com>
> wrote:
> >
> > I am currently working on a proposal where, if it gets funded, we will
> start to solve Maxwell's equations
> >
> > —
> > You are receiving this because you were mentioned.
>
> > Reply to this email directly, view it on GitHub, or mute the thread.
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/10238#issuecomment-373760540>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHPulDhRbXIYjFmb9ga7NXqxTKeBN45Aks5te-H9gaJpZM4QZtul>
> .
>

cticenhour commented:
I am currently working on implementing the Helmholtz equation for electric and magnetic fields in MOOSE as part of my dissertation research. One planned application for the code would be in RF wave heating (ICRH in fusion tokamaks) where we would be using a dielectric tensor to represent the plasma physics (but I haven't gotten there quite yet). @mangerij, I was actually initially thinking of using the materials system for defining this tensor since it seems to have much of what I would need to get started, at first glance.
